feugy/tabulous
# chore(web): improves webRTC connectivity

### What's in there?

Connecting with peers in a game is not super reliable yet.
These changes slightly improve the situation, without making it super stable.

Are included here:

- [feat(web): displays a label on host player avatar](https://github.com/feugy/tabulous/commit/86b2e7674007d29e2a70d91d9f2eb3c3475b5730)
- [refactor(web): maintain playing status at game level](https://github.com/feugy/tabulous/commit/e8261167451e7dfeda8fe3e122bdebfd0723e8e6)
- [refactor(server): removes game list update on playing status changes](https://github.com/feugy/tabulous/commit/d9b2f4337e4c0a5479fa36c0b3cbb8a6e46b3e09)
- [chore(web): populates host status when taking role](https://github.com/feugy/tabulous/commit/4487cec601d8768007c04e80eb6b3b9b2824388a)
- [chore(web): limits webRTC video bitrate for better experience](https://github.com/feugy/tabulous/commit/0ae17c59f00348dbd284df79604aca2bad6dfae8)
- chore(web): on vite reload, all players could become hosts or peers simultaneously

### How to test?

This needs 2 to 3 camera/mic to work
1. Make a game and invite two others
2. Let them join
   > Everyone should see 3 videos, one being tagged as "host" (ideally the same for all)
3. A peer refreshes their page
   > Video disappears, and re-appears. Host hasn't changed
4. Host leaves the game
   > Their video disappears, a new host is elected (the same for the 2 remaining players)
4. Former host joins the game
   > Their video appears, but they are not host

### Notes to reviewers

In this implementation, player's `playing` status is managed on server side when they establishing and closing their signal subscription.
However it is only used when client loads the game data, to determine the host.
After connecting with peers, client will instead rely on webRTC channel to know who's playing and who's not.
Turning host does not require a server round-trip any more, and only rely on these client-maintained statuses.

During Vite HMR, multiple client connected to the same game will all: either become host, either wait endlessly.
To fix this, HMR is delayed on non-host peers. However, the solution is not supported under Jest. A custom transformer has to replace `import.meta.hot` during code transformation to unlock testing.
